File: external/cnem_03-10-17/matlab/tuto3d/tuto_meca3d.m
% illoul_lounes@yahoo.fr
% ENSAM Paris

clear all;
close all; 
clc;

addpath('../bin');

%% load domain (nodes + boundary(s)) and  draw boundary(s) 

load('data_scni_mth_1'); 

figure;
tri_out_handle=trimesh(IN_Tri_Ini,XYZ_Noeud(:,1),XYZ_Noeud(:,2),XYZ_Noeud(:,3),'edgecolor','black');
alpha(tri_out_handle,0.5);
axis vis3d
axis equal

%% SCNI integration

Sup_NN_GS = 0;
Type_FF = 0;

[GS,XYZ_Noeud,IN_Tri_Ini,IN_Tri,IN_Tet,INV_NN,PNV_NN,IN_New_Old,IN_Old_New]=...
    m_cnem3d_scni(XYZ_Noeud,IN_Tri_Ini,Type_FF,Sup_NN_GS);

%% mechenical probelm : shearing cube with hole

%% Assembling

n=size(XYZ_Noeud,1);

E_NU=[200e9,0.266];

[B,C,V]= m_assembling_meca(GS,E_NU);
K_Mat_meca=B'*V*C*B;

%% boundary condition definition

Ind_Node_Front=unique(reshape(IN_Tri,1,[]));

UVW=zeros(3*n,1);
DDL_bloc=[];

VW_bh=0.; 
U_h=0.1;
U_b=0.; 

for i=1:size(Ind_Node_Front,2)
    Ind_i=Ind_Node_Front(i);
    N=XYZ_Noeud(Ind_i,:);
    
    if abs(N(1,3)-0.)<1.e-3
        DDL=3*(Ind_i-1)+1+0;
        DDL_bloc=[DDL_bloc,DDL];
        UVW(DDL)=U_b;
        DDL=3*(Ind_i-1)+1+1;
        DDL_bloc=[DDL_bloc,DDL];
        UVW(DDL)=VW_bh;
        DDL=3*(Ind_i-1)+1+2;
        DDL_bloc=[DDL_bloc,DDL];
        UVW(DDL)=VW_bh;
    elseif abs(N(1,3)-1.)<1.e-3
        DDL=3*(Ind_i-1)+1+0;
        DDL_bloc=[DDL_bloc,DDL];
        UVW(DDL)=U_h;
        DDL=3*(Ind_i-1)+1+1;
        DDL_bloc=[DDL_bloc,DDL];
        UVW(DDL)=VW_bh;
        DDL=3*(Ind_i-1)+1+2;
        DDL_bloc=[DDL_bloc,DDL];
        UVW(DDL)=VW_bh;
    end
end

DDL_bloc=unique(DDL_bloc);
DDL_Tot=1:3*n;
DDL_Libre=setdiff(DDL_Tot,DDL_bloc);

%% Solve problem

Vec_B=-K_Mat_meca(:,DDL_bloc)*UVW(DDL_bloc);

UVW(DDL_Libre)=K_Mat_meca(DDL_Libre,DDL_Libre)\Vec_B(DDL_Libre);

%% Post

Sig=C*(B*UVW);

SigMis=((Sig(1:6:size(Sig,1))-Sig(2:6:size(Sig,1))).^2+...
        (Sig(1:6:size(Sig,1))-Sig(3:6:size(Sig,1))).^2+...
        (Sig(2:6:size(Sig,1))-Sig(3:6:size(Sig,1))).^2+...
      6*(Sig(4:6:size(Sig,1)).^2+Sig(5:6:size(Sig,1)).^2+Sig(6:6:size(Sig,1)).^2));
SigMis=(SigMis./2.).^0.5;

UVW=reshape(UVW,3,n)';
Sig=reshape(Sig,6,n)';

%% print result (paraview)

out_var_vtk(XYZ_Noeud(1:n,:),IN_Tet,[Sig,SigMis],UVW,0);

%%

File: external/cnem_03-10-17/python/tuto/tuto_meca3d.py
##
## Copyright (C) 2003-2011
## Lounes ILLOUL (illoul_lounes@yahoo.fr)
## Philippe LORONG (philippe.lorong@ensam.eu)
## Arts et Metiers ParisTech, Paris, France
##
## $Revision: 25 $ - $Date: 2012-02-22 17:36:30 +0100 (mer., 22 fevr. 2012) $

import sys
import os
import shutil
import re
import math
import cPickle
import numpy as np
import scipy as sp
from scipy.sparse import coo_matrix
from scipy.sparse.linalg.dsolve import spsolve
from scipy.sparse.linalg.isolve import gmres

sys.path.append('../bin')

import load_node_front
import py_cnem3d
import check_gs
import py_assembling
import outvar

#-------------------------------------------------------------------------------
# read nodes and boundary :
print '\n--------------------------------------------------------------------------------\n'
print 'read node and boundary...\n'

[XYZ_Noeud,IN_Tri_Ini]=load_node_front.load3d('data3d_small/')

#-------------------------------------------------------------------------------
# call scni cnem3d :
print '\n--------------------------------------------------------------------------------\n'

Type_FF=0
Sup_NN_GS=0

[GS,XYZ_Noeud,IN_Tri_Ini,IN_Tri,IN_Tet,INV_NN,PNV_NN,INNO,INON]=\
py_cnem3d.scni(XYZ_Noeud,IN_Tri_Ini,Type_FF,Sup_NN_GS)

#print '\ncheck scni...\n'
#check_gs.check(GS,XYZ_Noeud[0:n,:])

#-------------------------------------------------------------------------------
# assembling for mechanic  :
print '\n--------------------------------------------------------------------------------\n'
print 'assembling...\n'

n=len(GS[0])

E_NU=[(200.e9,0.266)]

[B,C,V]=py_assembling.meca(GS,E_NU)
MK=B.transpose()*(V*C)*B

#-------------------------------------------------------------------------------
#solve case :
print '\n--------------------------------------------------------------------------------\n'
print 'solve case...\n'

#boundary condition definition :
#   U,V,W deplacement = 0 on Z=0   
#   U,W = 0 and V=0.1 on Z=1 

v=0.1

n_z_0=list()
n_z_1=list()

IN_Front=set(IN_Tri.flatten())

for i in IN_Front:
    N=XYZ_Noeud[i,:]
    
    if abs(N[2])<1.e-6:
        n_z_0.append(i)
    elif abs(N[2]-1.)<1.e-6:
        n_z_1.append(i)

n_z_0=np.array(n_z_0)
n_z_1=np.array(n_z_1)

DDL_U_z_0=n_z_0*3
DDL_V_z_0=n_z_0*3+1
DDL_W_z_0=n_z_0*3+2
DDL_U_z_1=n_z_1*3
DDL_V_z_1=n_z_1*3+1
DDL_W_z_1=n_z_1*3+2

UVW=np.zeros(3*n)

UVW[DDL_U_z_0]=0.
UVW[DDL_V_z_0]=0.
UVW[DDL_W_z_0]=0.
UVW[DDL_U_z_1]=0.
UVW[DDL_V_z_1]=v
UVW[DDL_W_z_1]=0.

DDL_bloc=np.hstack([DDL_U_z_0,DDL_V_z_0,DDL_W_z_0,DDL_V_z_1,DDL_U_z_1,DDL_W_z_1])
DDL_bloc=np.unique(DDL_bloc)
DDL_Tot=np.array(range(3*n))
DDL_Libre=np.setdiff1d(DDL_Tot,DDL_bloc)

b=-(MK[:,DDL_bloc]*UVW[DDL_bloc])
MK=MK[DDL_Libre,:]
MK=MK[:,DDL_Libre]

#solve the linear problem

UVW[DDL_Libre]=spsolve(MK,b[DDL_Libre])
#UVW[DDL_Libre]=gmres(MK,b[DDL_Libre])[0]

Strain=B*UVW 
Stress=C*Strain;
Stress=Stress.reshape(Stress.shape[0]/6,6)
UVW=UVW.reshape((UVW.shape[0]/3,3))

#-------------------------------------------------------------------------------
# write the result :
print '\n--------------------------------------------------------------------------------\n'
print 'write result to vtk format for paraview'

outvar.vtk(XYZ_Noeud,IN_Tet,Stress,UVW,0)
    
#-------------------------------------------------------------------------------

